### PR TITLE
[CP Staging] Feat/36985 create new rate field followups

### PR DESCRIPTION
**Pull request by koko57**
### Fixed Issues

$ https://github.com/Expensify/App/issues/36985
$ https://github.com/Expensify/App/issues/40712
$ https://github.com/Expensify/App/issues/40722
PROPOSAL: https://github.com/Expensify/App/issues/36985#issuecomment-2069183035

### Tests

Verify that changes don't break these areas:

**MoneyRequestConfirmationList**
1. Go to global create (+) > Submit expense
2. Change the request type to Distance
3. Choose and submit the start/finish points
4. Choose a participant and go to the Confirmation step
5. Verify that the distance is displayed correctly
6. Verify that clicking the distance field navigates to the distance editing page

**Rate highlighting**
1. Repeat steps 1-3 from above
2. Choose a workspace as a participant and go to the Confirmation step
3. Click on the Rate field
4. Verify that on the Rate selection page the chosen rate is also highlighted

**MoneyRequestView**
1. Go to the distance expense preview 
5. Verify that the rate field is displayed correctly with a proper rate

-> edit distance

1. Click Distance.
2. Change the distance and save it.
6. Verify that until the distance is updated the field displays "Pending..." text.

- [x] Verify that no errors appear in the JS console

### Offline tests

**MoneyRequestConfirmationList**
1. Go to global create (+) > Track expense
2. Change the request type to Distance
7. Choose and submit the start/finish points
8. Verify that the distance field displays "Pending..." text.

**MoneyRequestView**
1. Go to distance expense preview 
2. Verify that the distance field displays "Pending..." text.

### QA Steps

Verify that changes don't break these areas:

**MoneyRequestConfirmationList**
1. Go to global create (+) > Submit expense
2. Change the request type to Distance
4. Choose and submit the start/finish points
5. Choose a participant and go to the Confirmation step
6. Verify that the distance is displayed correctly
7. Verify that clicking the distance field navigates to the distance editing page

**Rate highlighting**
1. Repeat steps 1-3 from above
2. Choose a workspace as a participant and go to the Confirmation step
3. Click on the Rate field
4. Verify that on the Rate selection page the chosen rate is also highlighted

**MoneyRequestView**
1. Go to distance expense preview 
2. Verify that the rate field is displayed correctly with a proper rate

-> edit distance

1. Click Distance.
2. Change the distance and save it.
3. Verify that until the distance is updated the field displays "Pending..." text.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

Screenshot 2024-04-22 at 18 13 04

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

File: src/libs/DistanceRequestUtils.ts [hunk 1]
```diff
@@ -49,6 +37,7 @@ Onyx.connect({
 
 const METERS_TO_KM = 0.001; // 1 kilometer is 1000 meters
 const METERS_TO_MILES = 0.000621371; // There are approximately 0.000621371 miles in a meter
+const distanceMerchantRegex = /^[0-9.]+ \w+ @ (-|-\()?(\p{Sc}|\p{L}|\w){1,3} ?[0-9.]+\)? \/ \w+$/u;
```
Comment (neil-marcellini) [inline src/libs/DistanceRequestUtils.ts:40]: Please add a comment explaining what this does. Something like this.
```ts
        // Validates that the merchant for a distance request transaction is in the expected format.
        // <distance> <unit> @ <formatted rate with currency> / <unit>
        // Example: 1.45 miles @ $0.67 / mi
```

File: src/components/ReportActionItem/MoneyRequestView.tsx [hunk 2]
```diff
@@ -186,7 +188,7 @@ function MoneyRequestView({
 
     const currency = policy ? policy.outputCurrency : PolicyUtils.getPersonalPolicy()?.outputCurrency ?? CONST.CURRENCY.USD;
 
-    const mileageRate = TransactionUtils.isCustomUnitRateIDForP2P(transaction) ? DistanceRequestUtils.getRateForP2P(currency) : rates[rateID as string] ?? {};
+    const mileageRate = TransactionUtils.isCustomUnitRateIDForP2P(transaction) ? DistanceRequestUtils.getRateForP2P(currency) : distanceRates[rateID as string] ?? {};
```
Comment (neil-marcellini) [inline src/components/ReportActionItem/MoneyRequestView.tsx:191]: Please resolve [this consideration](https://github.com/Expensify/App/pull/38543#discussion_r1572686251) about when the mileage rate mapping has changed, and also add some manual test steps for it.

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
Looking good, a couple small things and one very important consideration. Also, I wouldn't expect editing distance on a tracked expense to work yet, since it's still in progress on the backend.

Would you please add some tests for verifying the merchant format with the regex? You can manually edit the merchant to whatever you want if you open the transaction in OldDot.

**Review by neil-marcellini — DISMISSED** [review]
Nice looks great, thank you. We're still working on the test steps as you said in Slack.

**Review by neil-marcellini — DISMISSED** [review]
Nice, and it's more simple now too

File: src/libs/DistanceRequestUtils.ts (hunk 1 above)
Comment (koko57) [reply src/libs/DistanceRequestUtils.ts]: @getusha no longer necessary - we decided to remove this

**Comment by koko57** [comment]
@neil-marcellini one more addition from https://github.com/Expensify/App/issues/40722

**Review by neil-marcellini — DISMISSED** [review]

**Comment by koko57** [comment]
@getusha  - my bad - I didn't updated the Tests section - from Neil: https://github.com/Expensify/App/pull/40711#pullrequestreview-2015605534
> Also, I wouldn't expect editing distance on a tracked expense to work yet, since it's still in progress on the backend.

File: src/components/ReportActionItem/MoneyRequestView.tsx [hunk 3]
```diff
@@ -290,6 +293,7 @@ function MoneyRequestView({
                     onPress={() => Navigation.navigate(ROUTES.MONEY_REQUEST_STEP_DISTANCE.getRoute(CONST.IOU.ACTION.EDIT, iouType, transaction?.transactionID ?? '', report.reportID))}
                 />
             </OfflineWithFeedback>
+            {/* TODO: correct the pending field action https://github.com/Expensify/App/issues/36988 (?) */}
```
Comment (amyevans) [inline src/components/ReportActionItem/MoneyRequestView.tsx:296]: Why the `(?)`? 😄

File: src/libs/DistanceRequestUtils.ts [hunk 4]
```diff
@@ -227,6 +211,13 @@ function getMileageRates(policy: OnyxEntry<Policy>): Record<string, MileageRate>
     return mileageRates;
 }
 
+/**
+ * Retrieves the rate and unit for a P2P distance expense for a given currency.
+ *
+ * @param currency
+ * @returns The rate and unit in RateAndUnit object.
+ */
+
```
Comment (amyevans) [inline src/libs/DistanceRequestUtils.ts:220]:
```suggestion
```

File: src/components/ReportActionItem/MoneyRequestView.tsx (hunk 3 above)
Comment (koko57) [reply src/components/ReportActionItem/MoneyRequestView.tsx:296]: I wasn't sure if it's a proper issue to link

**Comment by koko57** [comment]
@getusha yes, i will look at it, it was reported in https://github.com/Expensify/App/issues/40762

**Comment by koko57** [comment]
@getusha @neil-marcellini I would say that it's rather a backend issue - I get the default rate ID as a response from UpdateMoneyRequestDistance

**Comment by koko57** [comment]
@neil-marcellini @mountiny if [it could be possibly a BE](https://github.com/Expensify/App/pull/40711#issuecomment-2072082639) problem and editing tracked expense distance doesn't work yet, shouldn't we disable editing the distance for now?

**Comment by neil-marcellini** [comment]
@koko57 would  you please fix the issue reported [here](https://github.com/Expensify/App/pull/40711#issuecomment-2072082639) where the distance field doesn't show the optimistically updated distance?

**Comment by koko57** [comment]
@neil-marcellini it should be already fixed

File: src/components/ReportActionItem/MoneyRequestView.tsx (hunk 3 above)
Comment (neil-marcellini) [reply src/components/ReportActionItem/MoneyRequestView.tsx]: Oh yeah, it should be this issue I think. Next time you could ask us directly or look at the [P2P distance project board](https://github.com/orgs/Expensify/projects/106/views/3).
https://github.com/Expensify/App/issues/36987

**Comment by koko57** [comment]
@getusha it's when you're clicking on the distance field?

**Comment by koko57** [comment]
@getusha I cannot recreate

**Comment by koko57** [comment]
@getusha @neil-marcellini PR ready for (I hope) last round of review

**Review by getusha — APPROVED** [review]

**Review by neil-marcellini — DISMISSED** [review]
Great thanks!

File: src/components/ReportActionItem/MoneyRequestView.tsx [hunk 5]
```diff
@@ -283,13 +285,14 @@ function MoneyRequestView({
...
                     description={translate('common.distance')}
-                    title={distanceToDisplay}
+                    title={getPendingFieldAction('waypoints') ? translate('iou.fieldPending') : distanceToDisplay}
                     interactive={canEditDistance}
                     shouldShowRightIcon={canEditDistance}
                     titleStyle={styles.flex1}
                     onPress={() => Navigation.navigate(ROUTES.MONEY_REQUEST_STEP_DISTANCE.getRoute(CONST.IOU.ACTION.EDIT, iouType, transaction?.transactionID ?? '', report.reportID))}
                 />
             </OfflineWithFeedback>
+            {/* TODO: correct the pending field action https://github.com/Expensify/App/issues/36988 */}
```
Comment (amyevans) [inline src/components/ReportActionItem/MoneyRequestView.tsx:295]:
```suggestion
            {/* TODO: correct the pending field action https://github.com/Expensify/App/issues/36987 */}
```

NAB but I believe @neil-marcellini said it should be 36987

**Review by amyevans — DISMISSED** [review]
LGTM just 1 non-blocking question/comment

**Review by amyevans — APPROVED** [review]

**Pull request merged**